pkg/Cpanel/Security/Advisor/Assessors/Trojans.pm

```diff
@@ -73,7 +92,521 @@ sub _check_for_libkeyutils {
+    if ( $ssh_G !~ /illegal|unknown/ ) {
+        $self->add_bad_advice(
+            'text'       => [ "Trojan sshd binary check: ssh -G failed to return illegal/unknown indicating the possibility of the Ebury rootkit." ],
+            'suggestion' => [
+                'Check page 57 from the following for more information "[output,url,_1,We Live Security More Information,_2,_3]"',
+                'http://www.welivesecurity.com/wp-content/uploads/2014/03/operation_windigo.pdf',
+                'target',
+                '_blank'
+            ],
+        );
+    }
+}
+
+sub _check_for_ebury_ssh_banner {
+    my ($self) = @_;
+    my ( $host, $port, $ssh_banner );
+    my $ssh_connection = $ENV{'SSH_CONNECTION'};
+    return if !$ssh_connection;
+
+    if ( $ssh_connection =~ m{ \s (\d+\.\d+\.\d+\.\d+) \s (\d+) \z }xms ) {
+        ( $host, $port ) = ( $1, $2 );
+    }
+
+    return if !$host;
+    return if !$port;
+
+    my $sock = IO::Socket::INET->new(
+        PeerAddr => $host,
+        PeerPort => $port,
+        Proto    => 'tcp',
+        Timeout  => 5,
+    ) or return;
+
+    $ssh_banner = readline $sock;
+    close $sock;
+    return if !$ssh_banner;
+    chomp $ssh_banner;
+
+    if ( $ssh_banner =~ m{ \A SSH-2\.0-[0-9a-f]{22,46} }xms ) {
+        $self->add_bad_advice(
+            'text'       => [ "Trojan sshd binary check: The sshd banner matches known signatures from Ebury machines, indicating the existence of the Ebury rootkit." ],
+            'suggestion' => [
+                'Check the following for more information "[output,url,_1,We Live Security More Information,_2,_3]"',
+                'http://www.welivesecurity.com/2014/02/21/an-in-depth-analysis-of-linuxebury/',
+                'target',
+                '_blank'
+            ],
+        );
+    }
+}
+
+sub _check_for_ebury_ssh_shmem {
+    my ($self) = @_;
+    # As far as we know, sshd sholudn't be using shared memory at all, so any usage is a strong 
+    # sign of ebury.
+    return if !defined( $IPCS_REF->{root}{mp} );
+    my $PROCESS_REF;
+    for my $href ( @{ $IPCS_REF->{root}{mp} } ) {
+        my $shmid = $href->{shmid};
+        my $cpid  = $href->{cpid};
+        if ( $PROCESS_REF->{$cpid}{CMD} && $PROCESS_REF->{$cpid}{CMD} =~ m{ \A /usr/sbin/sshd \b }x ) {
+            $self->add_bad_advice(
+                'text'       => [ "sshd shared memory check: A shared memory segment created by sshd process exists." ],
+                'suggestion' => [
+                    'Check pages 33-34 from the following for more information "[output,url,_1,We Live Security More Information,_2,_3]"',
+                    'http://www.welivesecurity.com/wp-content/uploads/2014/03/operation_windigo.pdf',
+                    'target',
+                    '_blank'
+                ],
+            );
+        }
+    }
+}
+
+sub get_ipcs_hash {
+    my %hash;
+    my $header = 0;
+	my @res = Cpanel::SafeRun::Simple::saferun( 'ipcs', '-m', '-p' );
```

Comment:
Err.  That's interesting.  The original Trojans.pm module was using that 
too.  All I did was copy the original line used there.

However, I just added use Cpanel::SafeRun::Simple; to the top of Trojans.pm

Thanks,
Peter


On 06/30/2016 10:02 AM, brian m. carlson wrote:
>
> In pkg/Cpanel/Security/Advisor/Assessors/Trojans.pm 
> <https://github.com/CpanelInc/addon_securityadvisor/pull/37#discussion_r69149196>:
>
> > +                'text'       => [ "sshd shared memory check: A shared memory segment created by sshd process exists." ],
> > +                'suggestion' => [
> > +                    'Check pages 33-34 from the following for more information "[output,url,_1,We Live Security More Information,_2,_3]"',
> > +                    'http://www.welivesecurity.com/wp-content/uploads/2014/03/operation_windigo.pdf',
> > +                    'target',
> > +                    '_blank'
> > +                ],
> > +            );
> > +        }
> > +    }
> > +}
> > +
> > +sub get_ipcs_hash {
> > +    my %hash;
> > +    my $header = 0;
> > +	my @res = Cpanel::SafeRun::Simple::saferun( 'ipcs', '-m', '-p' );
>
> You're not using the Cpanel::SafeRun::Simple module, so this doesn't 
> compile correctly.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/CpanelInc/addon_securityadvisor/pull/37/files/c8280e2124274bf78c43d11ed22dfd3adec5352d#r69149196>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AGJ9b25ZZrqZ1jC2MQ46sjAp4jCOMsRhks5qQ9prgaJpZM4Iw7k3>.
>

-- 
Peter Elsner
Technical Analyst III
Migration Specialist
cPanel, Inc.